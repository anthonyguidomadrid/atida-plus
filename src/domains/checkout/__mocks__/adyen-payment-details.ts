export const adyenPaymentDetailsPayload = {
  details: {
    redirectResult: 'Ab02b4c0!BQABAgBqYNHM/549oKe/2MLKe0Ph'
  },
  customer: {
    customer_reference: 'PT--1234',
    email: 'atida@atida.com',
    name: 'Atida Mifarma',
    is_guest: false
  }
}

export const adyenPaymentDetailsResponseData = {
  additionalData: {
    paymentMethod: 'paywithgoogle'
  },
  amount: {
    currency: 'EUR',
    value: 1000
  },
  merchantReference: '3b32711f-a6c8-442d-94f8-2a943c10a9c7',
  pspReference: 'QV7XMK8G9QRXGN82',
  resultCode: 'Authorised'
}
