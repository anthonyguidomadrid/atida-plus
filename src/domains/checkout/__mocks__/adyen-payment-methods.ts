export const adyenPaymentMethodsPayload = {
  allowedPaymentMethods: [
    'paywithgoogle',
    'applepay',
    'klarna_account',
    'mbway',
    'multibanco',
    'card'
  ],
  amount: 2199,
  countryCode: 'ES',
  currency: 'EUR',
  shopperLocale: 'es_ES',
  shopperReference: 'some-ref'
}

export const adyenPaymentMethodsResponseData = {
  paymentMethods: [
    {
      configuration: {
        gatewayMerchantId: 'AtidaCompanyECOM',
        merchantId: '50'
      },
      details: [{ key: 'paywithgoogle.token', type: 'payWithGoogleToken' }],
      name: 'Google Pay',
      type: 'paywithgoogle'
    },
    {
      configuration: {
        merchantId: '000000000203419',
        merchantName: 'AtidaCompanyECOM_DEV'
      },
      details: [{ key: 'applepay.token', type: 'applePayToken' }],
      name: 'Apple Pay',
      type: 'applepay'
    },
    { name: 'Conta Klarna', type: 'klarna_account' },
    {
      name: 'MB WAY',
      type: 'mbway'
    },
    {
      name: 'Multibanco',
      type: 'multibanco'
    },
    {
      brands: ['visa', 'mc'],
      details: [
        {
          key: 'encryptedCardNumber',
          type: 'cardToken'
        },
        {
          key: 'encryptedSecurityCode',
          type: 'cardToken'
        },
        {
          key: 'encryptedExpiryMonth',
          type: 'cardToken'
        },
        {
          key: 'encryptedExpiryYear',
          type: 'cardToken'
        },
        {
          key: 'holderName',
          optional: true,
          type: 'text'
        }
      ],
      name: 'Cartão de crédito',
      type: 'scheme'
    }
  ]
}

export const adyenPaymentMethodsResponse = {
  data: adyenPaymentMethodsResponseData,
  allowedPaymentMethods: [
    'adyen_googlepay',
    'adyen_applepay',
    'adyen_klarna_payovertime',
    'adyen_mbway',
    'adyen_multibanco',
    'adyen_card'
  ]
}
