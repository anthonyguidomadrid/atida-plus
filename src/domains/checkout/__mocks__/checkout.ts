export const customerData = {
  customer_reference: 'PT--1234',
  email: 'customer@test.com',
  name: 'Customer Test',
  is_guest: false
}
