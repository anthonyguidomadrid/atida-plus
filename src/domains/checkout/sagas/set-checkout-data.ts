import { Customer } from './../../account/types'
import {
  selectCustomerDetails,
  selectIsLoggedIn
} from './../../account/selectors/customer'
import type { PayloadAction } from '@reduxjs/toolkit'
import { AxiosResponse } from 'axios'
import type { SagaIterator } from 'redux-saga'
import {
  all,
  call,
  getContext,
  put,
  delay,
  takeLatest,
  select
} from 'redux-saga/effects'
import { createClient, logger } from '~helpers'
import {
  setDataTrigger,
  setDataRequest,
  setDataSuccess,
  setDataFailure,
  setDataFulfill
} from '../slices'
import { SetCheckoutDataPayload, CheckoutData } from '../types'
import { API_RESPONSE } from '~config/constants/api-response'
import { getErrorMessage } from '~helpers/error'
import { parseDateOfBirthFormat } from '~domains/account'

function* updateApi(payload: SetCheckoutDataPayload) {
  const locale: string = yield getContext('locale')
  const client = createClient({ locale, addAnonymousCustomerUniqueId: true })
  const numberOfAttempts = 5
  for (let i = 0; i < numberOfAttempts; i++) {
    try {
      const apiResponse: AxiosResponse<CheckoutData> = yield call(
        client.post,
        '/api/checkout/set-data',
        {
          ...payload
        }
      )
      return apiResponse
    } catch (error) {
      if (
        (error as Error).message?.includes('500') ||
        (error as Error).message?.includes('502')
      ) {
        updateApi(payload)
        yield delay(2000 + i * 1000)
        if (i === numberOfAttempts - 1) {
          throw new Error(API_RESPONSE.request_failed_after_several_attempt)
        }
        if (!window.history?.state?.url?.includes('checkout')) {
          throw error
        }
      } else {
        throw error
      }
    }
  }
}

function* doSetCheckoutData({
  payload
}: PayloadAction<SetCheckoutDataPayload>) {
  try {
    /* 
      PLUS-6195 - When FeatureFlag.ACCOUNT_PHONE_TO_CUSTOMER_LEVEL is removed,
      remove customer
    */
    const customer: Customer = yield select(selectCustomerDetails)
    const isLoggedIn: Customer = yield select(selectIsLoggedIn)
    yield put(setDataRequest(payload))
    const { shouldSetLoading, ...restPayload } = payload
    const response: AxiosResponse<CheckoutData> = yield call(updateApi, {
      ...restPayload,
      ...(isLoggedIn &&
        customer &&
        Object.keys(restPayload).length && {
          customer: {
            ...customer,
            salutation: customer?.salutation ?? 'Ms',
            ...(customer.dateOfBirth && {
              dateOfBirth: parseDateOfBirthFormat(customer.dateOfBirth)
            })
          }
        })
    } as SetCheckoutDataPayload)

    yield put(setDataSuccess(response.data))
  } catch (error) {
    const errorMessages = getErrorMessage(error)
    logger.error(errorMessages)
    if (Array.isArray(errorMessages)) {
      yield put(
        setDataFailure({
          message: errorMessages[0]
        })
      )
    } else {
      yield put(
        setDataFailure({
          message: errorMessages
        })
      )
    }
  } finally {
    yield delay(0)
    yield put(setDataFulfill())
  }
}

export function* setCheckoutDataSaga(): SagaIterator {
  yield all([takeLatest(setDataTrigger, doSetCheckoutData)])
}
