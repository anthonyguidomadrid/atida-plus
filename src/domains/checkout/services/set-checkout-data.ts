import getConfig from 'next/config'
import { normalizeBasket } from '~domains/basket'
import { fetchProductsForBasket } from '~domains/basket/services/fetch-products-for-basket'
import { SprykerBasketItem } from '~domains/basket/types'
import { createClient } from '~helpers'
import { normalizeCheckoutData } from '../normalizers/checkout-data'
import type {
  CheckoutData,
  SprykerCheckoutData,
  SprykerIncludedBasket
} from '../types'

if (typeof window !== 'undefined')
  throw new Error('This file should not be called client-side')

export const setCheckoutData = async (
  locale: string,
  data: CheckoutData,
  user: {
    token?: string
  }
): Promise<CheckoutData> => {
  const { serverRuntimeConfig } = getConfig()

  const client = createClient({
    locale,
    options: {
      baseURL: serverRuntimeConfig.ecommerceHost[locale],
      headers: user.token
        ? {
            Authorization: `Bearer ${user.token}`
          }
        : undefined
    }
  })

  const response = await client.post<SprykerCheckoutData>('/checkout-data', {
    data: {
      type: 'checkout-data',
      attributes: {
        ...(data?.taxReference && { taxReference: data?.taxReference }),
        customer: data.customer,
        billingAddress: data.billingAddress,
        shippingAddress: data.deliveryAddress,
        shipment: data.deliveryMethod
          ? {
              idShipmentMethod: Number(data.deliveryMethod)
            }
          : null,
        paymentMethod: data?.paymentMethods,
        payments: data?.payments
      }
    }
  })

  return {
    ...normalizeCheckoutData(response.data, data),
    basket: await fetchProductsForBasket(
      locale,
      normalizeBasket({
        data: response.data.included.find(
          (item): item is SprykerIncludedBasket => item.type === 'carts'
        ),
        included: response.data.included.filter(
          (item): item is SprykerBasketItem => item.type === 'items'
        )
      })
    )
  }
}
