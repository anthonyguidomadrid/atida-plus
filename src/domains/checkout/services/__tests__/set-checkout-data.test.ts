/**
 * @jest-environment node
 */
import axios from 'axios'
import {
  sprykerCheckoutData,
  sprykerCheckoutDataWithBasketItems
} from '../../__mocks__/checkout-data'
import { createClient } from '~helpers'
import { setCheckoutData } from '../set-checkout-data'
// @ts-ignore
import { mget } from '@elastic/elasticsearch'
import { elasticsearchProducts } from '../../../basket/__mocks__/basket'

describe(setCheckoutData, () => {
  beforeEach(() => {
    mget.mockResolvedValue(elasticsearchProducts)
  })

  it('creates the client & passes the locale and token to spryker request', async () => {
    ;(axios.post as jest.Mock).mockResolvedValue({ data: sprykerCheckoutData })
    await setCheckoutData(
      'en-gb',
      { deliveryMethod: '7' },
      {
        token: 'some-user-token'
      }
    )
    expect(createClient).toHaveBeenCalledWith({
      locale: 'en-gb',
      options: {
        baseURL: 'spryker.co.uk',
        headers: {
          Authorization: `Bearer some-user-token`
        }
      }
    })
    expect(axios.post).toHaveBeenCalledWith('/checkout-data', {
      data: {
        type: 'checkout-data',
        attributes: {
          billingAddress: undefined,
          shippingAddress: undefined,
          customer: undefined,
          shipment: {
            idShipmentMethod: 7
          }
        }
      }
    })
  })

  it('returns the normalized checkout data response', async () => {
    ;(axios.post as jest.Mock).mockResolvedValue({ data: sprykerCheckoutData })
    const response = await setCheckoutData(
      'en-gb',
      { deliveryMethod: '7' },
      {
        token: 'some-user-token'
      }
    )
    expect(response).toMatchSnapshot()
  })

  it('returns the normalized checkout data with basket items response', async () => {
    ;(axios.post as jest.Mock).mockResolvedValue({
      data: sprykerCheckoutDataWithBasketItems
    })
    const response = await setCheckoutData(
      'en-gb',
      { deliveryMethod: '7' },
      {
        token: 'some-user-token'
      }
    )
    expect(response).toMatchSnapshot()
  })
})
