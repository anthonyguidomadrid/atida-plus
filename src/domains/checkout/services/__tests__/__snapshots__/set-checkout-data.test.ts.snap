// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setCheckoutData returns the normalized checkout data response 1`] = `
Object {
  "basket": Object {
    "discountedItems": Array [],
    "items": Array [],
    "promotionalItems": Array [],
  },
  "billingAddress": Object {
    "address1": "Avda. Málaga 2, Pueblo Andaluz, Bl5A",
    "address2": null,
    "address3": " ",
    "cellPhone": null,
    "city": "Portimao",
    "comment": null,
    "company": null,
    "country": "Spain",
    "description": null,
    "email": null,
    "firstName": "Test",
    "isDefaultBilling": true,
    "isDefaultShipping": true,
    "iso2Code": "PT",
    "lastName": "Last Name",
    "middleName": null,
    "phone": "68373739",
    "poBox": null,
    "salutation": "Mr",
    "zipCode": "8142",
  },
  "cartId": "1427e0dc-8550-5239-8e7e-1a6c32e8d52c",
  "deliveryAddress": Object {
    "address1": "Avda. Málaga 2, Pueblo Andaluz, Bl5A",
    "address2": null,
    "address3": " ",
    "cellPhone": null,
    "city": "Portimao",
    "comment": null,
    "company": null,
    "country": "Spain",
    "description": null,
    "email": null,
    "firstName": "Test",
    "isDefaultBilling": true,
    "isDefaultShipping": true,
    "iso2Code": "PT",
    "lastName": "Last Name",
    "middleName": null,
    "phone": "68373739",
    "poBox": null,
    "salutation": "Mr",
    "zipCode": "8142",
  },
  "deliveryMethod": "7",
  "lastUsedPaymentCode": "braintree",
  "maxDeliveryDays": 3,
  "minDeliveryDays": 1,
  "paymentMethods": Array [
    Object {
      "attributes": Object {
        "availableAmount": 10,
        "paymentMethodName": "Internal",
        "paymentProviderName": "AtidaPayment",
        "priority": null,
        "requiredRequestData": Array [
          "paymentMethod",
          "paymentProvider",
        ],
      },
      "id": "3",
      "type": "payment-methods",
    },
  ],
  "shipmentMethods": Array [
    Object {
      "attributes": Object {
        "carrierName": "Spryker Dummy Shipment",
        "currencyIsoCode": "EUR",
        "deliveryTime": null,
        "maxDeliveryDays": 3,
        "minDeliveryDays": 1,
        "name": "Standard",
        "price": 490,
      },
      "id": "7",
      "type": "shipment-methods",
    },
    Object {
      "attributes": Object {
        "carrierName": "Spryker Dummy Shipment",
        "currencyIsoCode": "EUR",
        "deliveryTime": null,
        "maxDeliveryDays": 3,
        "minDeliveryDays": 1,
        "name": "Express",
        "price": 590,
      },
      "id": "8",
      "type": "shipment-methods",
    },
  ],
  "taxReference": "12345678A",
}
`;

exports[`setCheckoutData returns the normalized checkout data with basket items response 1`] = `
Object {
  "basket": Object {
    "cartCoupons": Array [],
    "coupons": Array [],
    "currency": "EUR",
    "discountTotal": 0,
    "discountedItems": Array [],
    "discounts": Array [],
    "expenses": Array [
      Object {
        "quantity": 1,
        "sumPriceToPayAggregation": 399,
        "type": "SHIPMENT_EXPENSE",
        "unitGrossPrice": 0,
        "unitNetPrice": 0,
      },
    ],
    "grandTotal": 9999,
    "id": "f0849e9c-6712-54b9-8db5-440438d03387",
    "items": Array [
      Object {
        "id": "100000001",
        "product": Object {
          "alternativeUrls": Object {
            "es-es": "",
            "pt-pt": "forcapil-anti-caida-del-cabello-arkopharma-30-capsulas-2+1",
          },
          "availability": "AVAILABLE",
          "brand": Object {
            "code": "atida",
            "label": "Atida",
          },
          "categories": Object {
            "lvl0": Array [
              "main",
            ],
          },
          "contentSizeFactor": 1,
          "description": "<p><strong>Forcapil Anti-Caída del Cabello Arkopharma 30 Cápsulas 2+1</strong>
</p><p>¡Recupera la fuerza y vitalidad de tu cabello con Forcapil!
</p><p>Indicado para hombres y mujeres, este nutricosmético previene y combate la caída del pelo.
</p><p>Fórmula exclusivamente con ingredientes 100% de origen natural.
</p><p>Tiene una alta concentración en biotina, y su composición con aminoácidos, vitaminas, minerales y plantas favorece el crecimiento de la melena.
</p><p>Se recomienda tomar en momentos en los que sufrimos una mayor pérdida de cabello, causada por estrés, menopausia, tras el parto, cambios de estación...
</p><p><strong>Tratamiento para 3 meses: 2 + 1 de REGALO.</strong>
</p>",
          "enabled": true,
          "format": Object {
            "code": "shampooCode",
            "label": "Shampoo",
          },
          "gtin": "8428148462962",
          "id": "100000001",
          "images": Array [
            Object {
              "productDatImage": "https://sandbox-atida.bynder.com/transform/db2e62c0-2c32-4e50-bee7-1a7a421323f1/ean_img_8833772266123_0",
              "productHigh": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20High-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
              "productTileThumbnail": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Thumbnail-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
            },
          ],
          "labels": Array [
            Object {
              "label": "Free sponge",
              "type": "promotion",
            },
            Object {
              "label": "Outlet",
              "type": "pricing-tag-outlet",
            },
            Object {
              "label": "-11%",
              "type": "discount",
            },
          ],
          "largeImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20High-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "manufacturerName": "LOreal Deutschland",
          "maxQuantity": 50,
          "mediumImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Medium-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "metaDescription": "Meta description No contiene gluten. Apto para personas celíacas Meta description",
          "metaKeywords": "Meta keywords NaturGreen Algarroba Meta keywords",
          "metaTitle": "Meta title NaturGreen Algarroba a la Taza BIO 225gr Meta title",
          "name": "Test Product 1",
          "netWeight": Object {
            "amount": 100,
            "unit": "Grams",
            "unitLabel": "g",
          },
          "pricePerUnit": Object {
            "currency": "EUR",
            "unit": "100 ml",
            "value": 8888,
          },
          "productDatImage": "https://sandbox-atida.bynder.com/transform/db2e62c0-2c32-4e50-bee7-1a7a421323f1/ean_img_8833772266123_0",
          "productTileImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "productTileRetinaImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Retina-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "pzn": "1234567",
          "rating": Object {
            "averageRating": 3.5,
            "numberOfReviews": 100,
          },
          "rrp": Object {
            "currency": "EUR",
            "value": 9999,
          },
          "shortDescription": "Preparado vegetal con harina de algarroba ecológica, ideal para desayunos y meriendas",
          "thumbnailImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Thumbnail-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "unitVolume": Object {
            "amount": 100,
            "unit": "MILLILITER",
            "unitLabel": "ml",
          },
          "url": "/forcapil-anti-caida-del-cabello-arkopharma-30-capsulas-2+1",
        },
        "quantity": 1,
        "sku": "100000001",
        "subtotal": 9999,
        "unitPrice": 9999,
      },
      Object {
        "id": "100000002",
        "product": Object {
          "alternativeUrls": Object {
            "es-es": "",
            "pt-pt": "forcapil-anti-caida-del-cabello-arkopharma-30-capsulas-2+1",
          },
          "availability": "AVAILABLE",
          "brand": Object {
            "code": "atida",
            "label": "Atida",
          },
          "categories": Object {
            "lvl0": Array [
              "main",
            ],
          },
          "contentSizeFactor": 1,
          "description": "<p><strong>Forcapil Anti-Caída del Cabello Arkopharma 30 Cápsulas 2+1</strong>
</p><p>¡Recupera la fuerza y vitalidad de tu cabello con Forcapil!
</p><p>Indicado para hombres y mujeres, este nutricosmético previene y combate la caída del pelo.
</p><p>Fórmula exclusivamente con ingredientes 100% de origen natural.
</p><p>Tiene una alta concentración en biotina, y su composición con aminoácidos, vitaminas, minerales y plantas favorece el crecimiento de la melena.
</p><p>Se recomienda tomar en momentos en los que sufrimos una mayor pérdida de cabello, causada por estrés, menopausia, tras el parto, cambios de estación...
</p><p><strong>Tratamiento para 3 meses: 2 + 1 de REGALO.</strong>
</p>",
          "enabled": true,
          "format": Object {
            "code": "shampooCode",
            "label": "Shampoo",
          },
          "gtin": "8428148462962",
          "id": "100000002",
          "images": Array [
            Object {
              "productDatImage": "https://sandbox-atida.bynder.com/transform/db2e62c0-2c32-4e50-bee7-1a7a421323f1/ean_img_8833772266123_0",
              "productHigh": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20High-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
              "productTileThumbnail": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Thumbnail-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
            },
          ],
          "labels": Array [
            Object {
              "label": "Free sponge",
              "type": "promotion",
            },
            Object {
              "label": "Outlet",
              "type": "pricing-tag-outlet",
            },
            Object {
              "label": "-11%",
              "type": "discount",
            },
          ],
          "largeImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20High-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "manufacturerName": "LOreal Deutschland",
          "maxQuantity": 50,
          "mediumImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Medium-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "metaDescription": "Meta description No contiene gluten. Apto para personas celíacas Meta description",
          "metaKeywords": "Meta keywords NaturGreen Algarroba Meta keywords",
          "metaTitle": "Meta title NaturGreen Algarroba a la Taza BIO 225gr Meta title",
          "name": "Test Product 2",
          "netWeight": Object {
            "amount": 100,
            "unit": "Grams",
            "unitLabel": "g",
          },
          "pricePerUnit": Object {
            "currency": "EUR",
            "unit": "100 ml",
            "value": 8888,
          },
          "productDatImage": "https://sandbox-atida.bynder.com/transform/db2e62c0-2c32-4e50-bee7-1a7a421323f1/ean_img_8833772266123_0",
          "productTileImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "productTileRetinaImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Retina-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "pzn": "1234567",
          "rating": Object {
            "averageRating": 3.5,
            "numberOfReviews": 100,
          },
          "rrp": Object {
            "currency": "EUR",
            "value": 9999,
          },
          "shortDescription": "Preparado vegetal con harina de algarroba ecológica, ideal para desayunos y meriendas",
          "thumbnailImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Thumbnail-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "unitVolume": Object {
            "amount": 100,
            "unit": "MILLILITER",
            "unitLabel": "ml",
          },
          "url": "/forcapil-anti-caida-del-cabello-arkopharma-30-capsulas-2+1",
        },
        "quantity": 7,
        "sku": "100000002",
        "subtotal": 9999,
        "unitPrice": 9999,
      },
      Object {
        "id": "100000003",
        "product": Object {
          "alternativeUrls": Object {
            "es-es": "",
            "pt-pt": "forcapil-anti-caida-del-cabello-arkopharma-30-capsulas-2+1",
          },
          "availability": "AVAILABLE",
          "brand": Object {
            "code": "atida",
            "label": "Atida",
          },
          "categories": Object {
            "lvl0": Array [
              "main",
            ],
          },
          "contentSizeFactor": 1,
          "description": "<p><strong>Forcapil Anti-Caída del Cabello Arkopharma 30 Cápsulas 2+1</strong>
</p><p>¡Recupera la fuerza y vitalidad de tu cabello con Forcapil!
</p><p>Indicado para hombres y mujeres, este nutricosmético previene y combate la caída del pelo.
</p><p>Fórmula exclusivamente con ingredientes 100% de origen natural.
</p><p>Tiene una alta concentración en biotina, y su composición con aminoácidos, vitaminas, minerales y plantas favorece el crecimiento de la melena.
</p><p>Se recomienda tomar en momentos en los que sufrimos una mayor pérdida de cabello, causada por estrés, menopausia, tras el parto, cambios de estación...
</p><p><strong>Tratamiento para 3 meses: 2 + 1 de REGALO.</strong>
</p>",
          "enabled": true,
          "format": Object {
            "code": "shampooCode",
            "label": "Shampoo",
          },
          "gtin": "8428148462962",
          "id": "100000003",
          "images": Array [
            Object {
              "productDatImage": "https://sandbox-atida.bynder.com/transform/db2e62c0-2c32-4e50-bee7-1a7a421323f1/ean_img_8833772266123_0",
              "productHigh": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20High-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
              "productTileThumbnail": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Thumbnail-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
            },
          ],
          "labels": Array [
            Object {
              "label": "Free sponge",
              "type": "promotion",
            },
            Object {
              "label": "Outlet",
              "type": "pricing-tag-outlet",
            },
            Object {
              "label": "-11%",
              "type": "discount",
            },
          ],
          "largeImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20High-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "manufacturerName": "LOreal Deutschland",
          "maxQuantity": 50,
          "mediumImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Medium-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "metaDescription": "Meta description No contiene gluten. Apto para personas celíacas Meta description",
          "metaKeywords": "Meta keywords NaturGreen Algarroba Meta keywords",
          "metaTitle": "Meta title NaturGreen Algarroba a la Taza BIO 225gr Meta title",
          "name": "Test Product 3",
          "netWeight": Object {
            "amount": 100,
            "unit": "Grams",
            "unitLabel": "g",
          },
          "pricePerUnit": Object {
            "currency": "EUR",
            "unit": "100 ml",
            "value": 8888,
          },
          "productDatImage": "https://sandbox-atida.bynder.com/transform/db2e62c0-2c32-4e50-bee7-1a7a421323f1/ean_img_8833772266123_0",
          "productTileImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "productTileRetinaImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Retina-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "pzn": "1234567",
          "rating": Object {
            "averageRating": 3.5,
            "numberOfReviews": 100,
          },
          "rrp": Object {
            "currency": "EUR",
            "value": 9999,
          },
          "shortDescription": "Preparado vegetal con harina de algarroba ecológica, ideal para desayunos y meriendas",
          "thumbnailImage": "https://d2csxpduxe849s.cloudfront.net/media/07EADC8C-9435-4B3D-87CF5EDA882A36F2/F1974216-C01C-415C-B00F4AEBC6928147/Product%20Tile%20Thumbnail-51236DF8-2D7D-4757-AF2614FABA305C0A.png",
          "unitVolume": Object {
            "amount": 100,
            "unit": "MILLILITER",
            "unitLabel": "ml",
          },
          "url": "/forcapil-anti-caida-del-cabello-arkopharma-30-capsulas-2+1",
        },
        "quantity": 3,
        "sku": "100000003",
        "subtotal": 9999,
        "unitPrice": 9999,
      },
    ],
    "promotionalItems": Array [],
    "rrpTotal": 9999,
    "shippingTotal": 399,
    "subTotal": 9999,
    "taxTotal": 1596,
  },
  "billingAddress": Object {
    "address1": "Avda. Málaga 2, Pueblo Andaluz, Bl5A",
    "address2": null,
    "address3": " ",
    "cellPhone": null,
    "city": "Portimao",
    "comment": null,
    "company": null,
    "country": "Spain",
    "description": null,
    "email": null,
    "firstName": "Test",
    "isDefaultBilling": true,
    "isDefaultShipping": true,
    "iso2Code": "PT",
    "lastName": "Last Name",
    "middleName": null,
    "phone": "68373739",
    "poBox": null,
    "salutation": "Mr",
    "zipCode": "8142",
  },
  "cartId": "1427e0dc-8550-5239-8e7e-1a6c32e8d52c",
  "deliveryAddress": Object {
    "address1": "Avda. Málaga 2, Pueblo Andaluz, Bl5A",
    "address2": null,
    "address3": " ",
    "cellPhone": null,
    "city": "Portimao",
    "comment": null,
    "company": null,
    "country": "Spain",
    "description": null,
    "email": null,
    "firstName": "Test",
    "isDefaultBilling": true,
    "isDefaultShipping": true,
    "iso2Code": "PT",
    "lastName": "Last Name",
    "middleName": null,
    "phone": "68373739",
    "poBox": null,
    "salutation": "Mr",
    "zipCode": "8142",
  },
  "deliveryMethod": "7",
  "lastUsedPaymentCode": "braintree",
  "maxDeliveryDays": 3,
  "minDeliveryDays": 1,
  "paymentMethods": Array [
    Object {
      "attributes": Object {
        "availableAmount": 10,
        "paymentMethodName": "Internal",
        "paymentProviderName": "AtidaPayment",
        "priority": null,
        "requiredRequestData": Array [
          "paymentMethod",
          "paymentProvider",
        ],
      },
      "id": "3",
      "type": "payment-methods",
    },
  ],
  "shipmentMethods": Array [
    Object {
      "attributes": Object {
        "carrierName": "Spryker Dummy Shipment",
        "currencyIsoCode": "EUR",
        "deliveryTime": null,
        "maxDeliveryDays": 3,
        "minDeliveryDays": 1,
        "name": "Standard",
        "price": 490,
      },
      "id": "7",
      "type": "shipment-methods",
    },
    Object {
      "attributes": Object {
        "carrierName": "Spryker Dummy Shipment",
        "currencyIsoCode": "EUR",
        "deliveryTime": null,
        "maxDeliveryDays": 3,
        "minDeliveryDays": 1,
        "name": "Express",
        "price": 590,
      },
      "id": "8",
      "type": "shipment-methods",
    },
  ],
  "taxReference": "12345678A",
}
`;
